1-Draft/RFCnnnn-ForEach-Parallel-Cmdlet.md

```diff
@@ -16,68 +16,95 @@ This RFC proposes a new parameter set for the existing ForEach-Object cmdlet to
-Without a `-Timeout` parameter, the cmdlet will wait indefinitely for completion.
+Implementation will be similar to the ThreadJob module in that thread script block execution will be contained within a PSThreadChildJob object.
+The jobs will be run concurrently on separate runspaces/threads up to the ThrottleLimit value, and the remainder queued to wait for an available runspace/thread to run on.  
+Initial implementation will not attempt to reuse threads and runspaces when running queued items, due to concerns of stale state breaking script execution.
+For example, PowerShell uses thread local storage to store per thread default runspaces.
+And even though there is a runspace 'ResetRunspaceState' API method, it only resets session variables and debug/transaction managers.
+Imported modules and function definitions are not affected.
+A script that defines a constant function would fail if the function is already defined.  
+The initial assumption will be that runspace/thread creation time is insignificant compared to the time needed to execute the script block, either because of high compute needs or because of long wait times for results.
+If this assumption is not true then the user should consider batching the work load to each foreach-object iteration, or simply use the sequential/non-parallel form of the cmdlet.
 
-### Synchronous parameter set
+The 'TimeoutSeconds' parameter will attempt to halt all script block executions after the timeout time has passed, however it may not be immediately successful if the running script is calling a native command or API, in which case it needs for the call to return before it can halt the running script.
 
-Synchronous ForEach-Object -Parallel returns after all script blocks complete running or timeout
+### Variable passing
 
-```powershell
-ForEach-Object -Parallel -ThrottleLimit 10 -TimeoutSecs 1800 -ScriptBlock {} 
-```
+ForEach-Object -Parallel will support the PowerShell `$_` current piped item variable within each script block.
+It will also support the `$using:` directive for passing variables from script scope into the parallel executed script block scope.  
+If the passed in variable is a value type, a copy of the value is passed to the script block.  
+If the passed in variable is a reference type, the reference is passed and each running script block can modify it.
+Since the script blocks are running in different threads, modifying a reference type that is not thread safe will result in undefined behavior.  
 
-- `-Parallel` : parameter switch specifies fan-out parallel script block execution
+Script block variables will be special cased because they have runspace affinity.
+Therefore script block variables will not be passed by reference and instead a new script block object instance will be created from the original script block variable Ast (abstract syntax tree).

```

Comment:
Will the creation of a new scriptblock apply only during `$using:` and/or parameter passing?  Or will it also occur when the runspace encounters a scriptblock from another runspace at large?  An example of such an encounter I'm thinking of is

```PowerShell
$g = Get-SomeScriptBlockGenerator

1..10 | ForEach-Object -Parallel { . $using:g.GetNextScriptBlock() }
```

where the scriptblock produced by $g.GetNextScriptBlock() has affinity to the runspace where Get-SomeScriptBlockGenerator is invoked.

Reply:
Actually, this has changed.  Since it is impossible to cover all cases where script block variables can be passed to the foreach script block, it will not be supported (i.e., no attempt will be made to detect and pass script blocks via Ast or string).  An attempt will be made to catch this and error out when script block variables are passed directly through piped input or using variables, but not all possible variations will be checked (since it is not really possible).

Documentation will warn users not to do this, along with other unsafe uses of variables while running in parallel, such as assigning to passed in variables that are not thread safe.